### PR TITLE
WIP: Handle circular dependencies

### PR DESCRIPTION
**Pull request by KamranAsif**
Adding code to handle circular dependencies

### CODE REVIEW

File: test/revise-spec.ts [hunk 1]
```diff
@@ -28,75 +28,69 @@ describe('revise', () => {
     });
   });
 
-  describe('works with objects', () => {
+  describe('when objects passed in', () => {
     it('should return unchanged object if no change', () => {
-      const source = {foo: 1};
-      const revision = {foo: 1};
+      const source = {foo : 1};
```
Comment (wcjordan) [inline test/revise-spec.ts]: is this the recommended style?

seems bizarre to me

**Comment by wcjordan** [comment]
whats up w/ the test failures?

File: test/revise-spec.ts (hunk 1 above)
Comment (KamranAsif) [reply test/revise-spec.ts]: No, I had an issue with my clang-formatter. See #19

**Comment by KamranAsif** [comment]
@wcjordan Going for a TDD approach for this one. Not committing until I have the fixes in.

File: test/revise-spec.ts [hunk 2]
```diff
@@ -151,4 +151,64 @@ describe('revise', () => {
...
+    it('handles reversed pointing circular references', () => {
+      const source: any = {};
+      const revision: any = {};
+      source.foo = revision;
+      revision.foo = source;
+
+      const newSource = revise(source, revision);
+
+      // Strange case, it ends up being source..
+      assert.strictEqual(newSource.foo, revision);
```
Comment (KamranAsif) [inline test/revise-spec.ts]: @wcjordan This assert ended up changing from 
assert.strictEqual(newSource.foo, source); 
to 
assert.strictEqual(newSource.foo, revision);

Its strange but I think the logic works out?

**Comment by KamranAsif** [comment]
@wcjordan please review with ?w=1 url param

**Comment by KamranAsif** [comment]
Looks like this CL has performance issues. Going to investigate. Probably need to avoid WeakSet

File: test/revise-spec.ts (hunk 2 above)
Comment (wcjordan) [reply test/revise-spec.ts]: Ya the idea is we retain the source object right?  so the original source.foo is revision.

**Pull request merged**